lib/api/apiUtils/object/versioning.js

```diff
@@ -0,0 +1,161 @@
+import { errors, versioning } from 'arsenal';
+
+import metadata from '../../../metadata/wrapper';
+
+const versionIdUtils = versioning.VersionID;
+const VID_INF = versionIdUtils.VID_INF;
+
+export function decryptVersionId(reqQuery) {
+    if (!reqQuery || !reqQuery.versionId) {
+        return undefined;
+    }
+    // set to undefined if value is ''
+    let versionId = reqQuery.versionId || undefined;
+    if (!versionId || versionId === 'null') {
+        return versionId;
+    }
+    try {
+        versionId = versionIdUtils.decrypt(versionId);
+    } catch (exception) {
+        return errors.InvalidArgument
+            .customizeDescription('Invalid version id specified');
+    }
+    return versionId;
+}
+
+export function getVersionIdResHeader(verCfg, objectMD) {
+    if (verCfg) {
+        if (objectMD.isNull || (objectMD && !objectMD.versionId)) {
+            return 'null';
+        }
+        return versionIdUtils.encrypt(objectMD.versionId);
+    }
+    return undefined;
+}
+
+export function versioningPreprocessing(bucketName, bucketMD, objectKey,
+    objMD, reqVersionId, log, callback) {
+    const options = {};
+    // bucket is not versioning enabled
+    if (!bucketMD.getVersioningConfiguration()) {
+        options.deleteData = true;
+        return callback(null, options);
+    }
+    // bucket is versioning enabled
+    const mstVersionId = objMD ? objMD.versionId : undefined;
+    const mstIsNull = objMD ? objMD.isNull : false;
+    const vstat = bucketMD.getVersioningConfiguration().Status;
+    if (!reqVersionId) {
+        // non-version-specific versioning operation
+        if (mstVersionId === undefined || mstIsNull) {
+            // object does not exist or is not versioned (before versioning)
+            if (vstat === 'Suspended') {
+                // versioning is suspended, overwrite the existing version
+                options.versionId = '';
+                options.isNull = true;
+                options.deleteData = true;
+                return callback(null, options);
+            }
+            // versioning is enabled, create a new version
+            options.versioning = true;
+            if (objMD) {
+                // store master version in a new key
+                const versionId = mstIsNull ? mstVersionId : VID_INF;
+                objMD.versionId = versionId; // eslint-disable-line
+                objMD.isNull = true; // eslint-disable-line
+                options.nullVersionId = versionId;
+                return metadata.putObjectMD(bucketName, objectKey, objMD,
```

Comment:
please add a log line for the error in the callback so we know that the error is from storing the null mater version in a new key in metadata.